### PR TITLE
feat(integrations): add StartSessionPoller adapters for github + slack

### LINKED ISSUE #44: feat(integrations): add StartSessionPoller adapters for github + slack
> This issue body is the canonical chain document. Two ways to drive it:
>
> - **Claude-only (light loop):** `/fix-issue <this-issue-url>` — no codex turns.
> - **Hybrid codex+claude (full loop):** save this body as `.claude/orchestration/<wave>/<slug>.md` and run `pnpm cli orchestration run @<path>` per `.claude/commands/codex.md`. The H1 + Phase headings below conform to `docs/orchestration/chain-format.md`.

# feat(integrations): add StartSessionPoller adapters for github + slack

> First of three "v1-wire" chains. Adds adapter closures that wrap the existing
> ResultAsync-returning pollers in `packages/integrations/{github,slack}` into
> the `StartSessionPoller` contract that `createStartSessionTool` expects. Lands
> independently; PR 2 (#45) consumes the exports.

## Goal

Build two factory functions — `createGithubPoller` and `createSlackPoller` —
that each return a `StartSessionPoller` closure capturing the auth token.
The closures call the underlying poll functions and surface failures as
rejected promises (the `StartSessionPoller` contract is `Promise<void>` with
no error channel, so the boundary translation happens here via neverthrow's
`_unsafeUnwrap()`).

Out of scope: refactoring the `StartSessionPoller` contract to return
`Result`. That would touch the tool dispatcher, the test surface, and the
type plumbed through `apps/mcp-local`; it deserves a separate decision-record
issue and is explicitly NOT part of this chain.

## Read This First

- `CLAUDE.md` — stack, dev principles, package boundaries.
- `.claude/rules/workflow.md` — worktree convention, branch naming, "always
  merge never rebase," `pnpm validate` before push.
- `.claude/rules/pr-conventions.md` — conventional-commit title, ≤500 line
  diff, link the issue with `closes #44`.
- `.claude/rules/testing.md` — Polly cassettes via `definePollySetup`; no
  `.skip`/`.only`; cassettes in `src/__recordings__/`; never synthetic stubs
  in cassette tests.
- `.claude/rules/error-handling.md` — service-boundary scan covers
  `packages/integrations/{github,slack}/src/**`; no literal `throw`.
  `_unsafeUnwrap()` is a neverthrow API that rejects the promise on `Err` —
  this is NOT a literal `throw` and does not trip
  `pnpm cli check-service-boundaries`. Call this out in the PR description so
  it's visible to the reviewer.
- `.claude/rules/typescript-patterns.md` — named object params, no `any`,
  explicit return types on exported functions.
- `docs/CODEBASE_MAP.md` — particularly the integration-package section and
  the `StartSessionPoller` contract documentation.
- `packages/mcp-server/src/tools/composite/start-session.ts:60` — the
  `StartSessionPoller` type definition.
- `packages/integrations/github/src/polling.ts:56-64` — `pollPullRequests` /
  `pollIssues` / `pollMentions` signatures.
- `packages/integrations/slack/src/{mentions,dms}.ts` — Slack pollers.
- `packages/integrations/core/src/test-setup/polly.ts` — `definePollySetup`.

### In scope

- New file `packages/integrations/github/src/poller.ts` exporting
  `createGithubPoller({ token, username }): StartSessionPoller`. The
  closure:
  - Reads the current cursor from `integration_state` for each underlying
    sub-poll (the underlying pollers already do this via their `since`
    argument; thread `since` through from the cursor).
  - Calls `pollPullRequests` → `pollIssues` → `pollMentions` sequentially.
    Mentions requires `username` — passed in to the factory so the closure
    is dependency-free of network calls outside its body.
  - On `Err` from any sub-poll, `_unsafeUnwrap()` to reject the promise.
    Caller (PR 3) catches.
- New file `packages/integrations/slack/src/poller.ts` exporting
  `createSlackPoller({ token }): StartSessionPoller`. Same shape but
  - Calls `pollMentions` → `pollDMs` sequentially.
  - No pre-fetched identity — Slack pollers call `auth.test()` internally.
- Re-export from `packages/integrations/github/src/index.ts` and
  `packages/integrations/slack/src/index.ts`.
- Polly-cassette tests at `packages/integrations/github/src/poller.test.ts`
  and `packages/integrations/slack/src/poller.test.ts`.
  - Tests assert: closure invocation triggers underlying poll functions,
    `evidence_log` rows appear after invocation, `integration_state.cursor`
    advances.
  - Use existing `definePollySetup` infrastructure; cassettes recorded once
    via `POLLY_MODE=record pnpm test --filter <pkg>`.
- All five existing CI gates remain green. `check-service-boundaries`
  remains green (the sixth gate).

### Explicitly out of scope

- Refactoring `StartSessionPoller` to return `Result<void, BaseError>`.
  Separate decision-record issue; do NOT bundle.
- Aggregating partial-success state across sub-pollers within a single
  integration (e.g. Slack mentions succeeds but DMs fails). Throw on first
  failure; PR 3 catches at the integration level.
- Adding new error variants to `GithubError` or `SlackError`. Reuse existing
  ones; if the underlying poll's error suffices, that's correct.
- Updating `apps/mcp-local/src/cli.ts`. That's PR 2 (#45); this PR adds the
  exports only, leaving them unconsumed.

### Non-negotiable constraints

- Public repo. No real tokens, real workspace identifiers, real customer
  names, or anything that could be confused for one appears in code,
  comments, fixtures, or cassettes. Cassettes recorded against the
  maintainer's own account are redacted by the existing
  `definePollySetup` chokepoint; verify the diff before committing.
- The `pnpm cli check-service-boundaries` scanner must remain green. If
  the implementer feels tempted to add a literal `throw` inside the
  closures, stop — use `_unsafeUnwrap()` instead, or `.mapErr()` to a
  domain error and re-throw via Promise rejection through neverthrow's
  API. No `throw` keyword in scanner-covered files.
- Named object params for the factory functions (`.claude/rules/typescript-patterns.md`).
  The `StartSessionPoller` closure signature is fixed by the contract at
  `start-session.ts:60` (positional `args`), so the closure body uses that
  shape — the factory takes object params.
- Cassette files committed under `packages/integrations/{github,slack}/src/__recordings__/`.
  `.gitignore` allow-list already covers these paths. Verify with `git check-ignore -v <file>`.
- Conventional-commit PR title exactly: `feat(integrations): add StartSessionPoller adapters for github + slack`.
- Draft PR initially; convert to ready-for-review after `pnpm validate` green.
- `closes #44` in the PR body.

## Variables

- `{worktree}`: `v1-wire-poller-adapters`
- `{pr_url}`: GitHub PR URL (filled in after PR creation)

#### Step 1: Initial Plan (codex-plan)

```prompt
Plan the implementation of StartSessionPoller adapters for the github and
slack integration packages in slopweaver/slopweaver. The chain doc this
turn came from is the source of truth for scope; do not expand beyond it.

Investigate first, plan second:

- Read packages/mcp-server/src/tools/composite/start-session.ts:60 and
  confirm the exact StartSessionPoller signature. Confirm it's
  Promise<void> with no error channel. Note line 64 — pollers map type.
- Read packages/integrations/github/src/polling.ts in full. List the
  signatures of pollPullRequests, pollIssues, pollMentions. Confirm
  return types are ResultAsync<{ fetched, newCursor }, GithubError>.
  Note which require `since` (cursor) and which require additional
  fields (e.g. mentions wants `username`).
- Read packages/integrations/github/src/identity.ts. Find `fetchIdentity`
  (or equivalent) — what it returns, where it lives, whether it's
  exported. The PR 2 wire-up needs to call it; this PR doesn't, but the
  factory signature should expect a pre-resolved username so the factory
  itself is network-free.
- Read packages/integrations/slack/src/mentions.ts and
  packages/integrations/slack/src/dms.ts. Confirm Slack pollers call
  auth.test() internally (no pre-fetch needed). List required and
  optional args.
- Read packages/integrations-core/src/test-setup/polly.ts. Understand
  definePollySetup's redactor wiring. Note the per-package
  extraRedactors pattern.
- Read packages/integrations/github/src/polling.test.ts and any
  existing slack test to understand the cassette layout convention.
  Cassettes go under src/__recordings__/<suite>/<test>/recording.har.
- Sweep .claude/rules/error-handling.md for the _unsafeUnwrap policy.
  Confirm that _unsafeUnwrap is NOT flagged by check-service-boundaries
  (the scanner only flags literal `throw` statements). Note the policy
  on Carve-outs (throws that are allowed) section.

Then produce a file-by-file change plan listing exactly:

- New files (paths, named exports, full signatures).
- Re-export additions to src/index.ts.
- Test files (Polly setup, assertions, cassette names).
- The exact StartSessionPoller closure body for each adapter — pseudocode
  is fine, but make the control flow explicit (sub-poll order,
  _unsafeUnwrap placement, Promise.all vs sequential).
- A verification section: how to prove the closures invoke the underlying
  polls, how to prove evidence_log rows land, how to prove
  integration_state.cursor advances.

Surface every risk: cassette redaction, contract drift if the underlying
poller signatures change later, the implications of _unsafeUnwrap on
upstream catch handling.

Do NOT write code yet.
```

#### Step 2: Tighten and Finalise (codex-send)

```prompt
Reduce the plan to the smallest straight-line set of edits. For each
item: keep, drop, or defer to a follow-up issue. Specifically:

- Confirm the test pattern: one happy-path cassette per integration is
  sufficient for v1. Edge-case cassettes (404, 401, 429) can be follow-ups.
- Confirm the sequential sub-poll order (PRs → Issues → Mentions for
  GitHub; Mentions → DMs for Slack) is correct. Justify against
  rate-limit risk and freshness ordering.
- Confirm the _unsafeUnwrap policy is correct. Reject any plan that
  uses a literal `throw` inside the scanner-covered closures.
- Confirm the factory does not perform network I/O — the username for
  GitHub is passed in, fetchIdentity is the caller's responsibility (PR 2).
- Confirm cassette redaction is verified before commit; the
  definePollySetup chokepoint covers standard auth, but any new PII
  shape may need extraRedactors. List any that need adding.

Output: the final straight-line plan as a numbered list of edits, in
the order they should be applied. Implementer follows verbatim.
```

#### Step 3: Handoff to Claude

The runner builds the handoff preamble automatically. Claude receives the
final plan from Step 2 and applies it without modification.

#### Step 4: Claude Implements

Claude executes the plan in order:

1. Create `packages/integrations/github/src/poller.ts` with
   `createGithubPoller`.
2. Create `packages/integrations/slack/src/poller.ts` with
   `createSlackPoller`.
3. Add re-exports in both `index.ts` files.
4. Write Polly cassette tests in both `poller.test.ts` files.
5. Record cassettes once: `POLLY_MODE=record pnpm test --filter @slopweaver/integrations-github`
   then the slack equivalent. Skim cassette diffs for tokens/PII before
   staging.
6. Run the full local verification loop:
   `pnpm cli check-service-boundaries && pnpm format:check && pnpm lint && pnpm compile && pnpm test && pnpm knip`.
   All six gates must pass.
7. Commit using conventional-commits style:
   `feat(integrations): add StartSessionPoller adapters for github + slack`.
8. Push, open the PR with body referencing `closes #44`, explain the
   `_unsafeUnwrap` boundary translation, link to error-handling.md
   carve-out, attach the four-check verification.

#### Step 5: PR Review (codex-review)

```prompt
Review {pr_url}. This PR adds StartSessionPoller adapters for github
and slack in the slopweaver public repo. Sub-issue #44; part of the
v1-wire chain (3 PRs total).

Confirm specifically:

- The new files at packages/integrations/{github,slack}/src/poller.ts
  contain factory functions with named object params, explicit return
  types, no `any`, no literal `throw`.
- Underlying poller invocation order matches the plan (sequential, in
  the order specified). No Promise.all of sub-polls within a single
  integration (rate-limit reason).
- Error handling uses neverthrow's _unsafeUnwrap — NOT a literal `throw`.
  Verify by grepping the new files: `grep "throw " packages/integrations/{github,slack}/src/poller.ts`
  should return zero matches.
- The factory is network-free: fetchIdentity is NOT called inside
  createGithubPoller. Username is a constructor argument.
- Polly cassettes at src/__recordings__/poller/*.har are present and
  redacted (no real tokens, real workspace IDs, real account names).
- Test assertions are strong: evidence_log rows asserted with exact
  values, integration_state.cursor asserted with toBe (not toBeTruthy).
- No .skip / .only / .todo in any new test file.
- pnpm validate runs all six gates green.
- PR title is exactly the conventional-commit shape from the chain.
- PR body links closes #44 and explains the _unsafeUnwrap boundary
  translation visibly enough that the reviewer can find it.
- Diff size ≤500 lines excluding cassettes and lockfile.

Flag any drift from the chain's locked scope. Specifically reject any
attempt to refactor StartSessionPoller to return Result — that's
explicitly out of scope. Reject any literal `throw` in the new files.
Reject any new error variants added to GithubError/SlackError.

Reply LGTM - ready for local testing. when there are no blocking
findings.
```

#### Step 6: Fix Loop

If the reviewer surfaces P0/P1 findings, the runner hands them to the
implementer as a fix prompt and pushes a follow-up commit. Loop repeats
until the reviewer returns the success sentinel.

### Phase 4: Human Review and Manual QA

After CI green, runner halts at `awaiting_manual_qa`. Maintainer:

1. Reads the diff with public-repo eyes — confirm cassettes are scrubbed.
2. Verifies the new exports type-check at the call site (in a scratch
   file: `import { createGithubPoller } from '@slopweaver/integrations-github'`
   then check the inferred type via a typed call).
3. Squash-merges via the GitHub UI.
4. Worktree cleanup:
   ```bash
   git worktree remove ~/dev/worktrees/v1-wire-poller-adapters
   git branch -d worktree/v1-wire-poller-adapters
   ```

refs #2 · closes #44

### PR DESCRIPTION
**Pull request by lachiejames**
## What this PR does

Adds `createGithubPoller({ token, username })` and `createSlackPoller({ token })` factories in `packages/integrations/{github,slack}/src/poller.ts`. Each returns a `StartSessionPoller` closure (from `@slopweaver/mcp-server`) that reads the current cursor from `integration_state` **before each sub-poll** and threads it as `since`:

- **github** → `pollPullRequests` → `pollIssues` → `pollMentions` (sequential; `username` is captured by the factory)
- **slack** → `pollMentions` → `pollDMs` (sequential; identity is resolved per-sub-poll via `auth.test()`)

Sub-polls run sequentially (not `Promise.all`) so they share each platform's rate-limit budget without bursting; the order also mirrors v1 ranking weight. Boundary translation from neverthrow `Result` to the `Promise<void>` contract happens via `_unsafeUnwrap()`.

The factories are re-exported from each package's public entry; no consumer is wired up yet (that's PR 2 / #45).

## Why

Lays the groundwork for live polling in the `start_session` composite tool — first PR in the v1-wire chain.

closes #44 · refs #2

## Test plan

- [x] Polly cassette tests at `packages/integrations/{github,slack}/src/poller.test.ts` exercise the closure end-to-end against committed cassettes under `src/__recordings__/createGithubPoller-cassette_*` / `createSlackPoller-cassette_*`. Each suite asserts:
  - First invocation: `evidence_log` rows land with the correct kind-prefixed `external_id`; `integration_state` row is bracketed with `started`/`completed`/`cursor` (cursor is a string).
  - Second invocation: prior cursor is read and threaded as `since`, so the API returns 0 *new* items; no duplicate rows; cursor is **preserved** (not regressed to null); `integration_state` is re-bracketed.
- [x] Cassettes recorded with `POLLY_MODE=record pnpm test --filter @slopweaver/integrations-{github,slack}` against the maintainer's accounts. Redaction audited: github responses contain only public-repo data (scoped via `maybeScopeGithubSearchUrl`); slack responses have all message text, channel names, usernames, profile fields, and workspace URLs scrubbed by `slackRedactors`. No sensitive headers (`authorization` / `cookie` / `set-cookie`) survive the shared `definePollySetup` redaction.
- [x] `pnpm validate` green locally (all six gates: `check-service-boundaries`, `format:check`, `lint`, `compile`, `test`, `knip`).
- [x] CI green (will trigger on push).

## Notes for the reviewer

- **Cursor-re-read fix.** The first push of this PR read the cursor **once** at closure entry and threaded the same `since: null` to all three sub-polls. Recording fresh cassettes against my real GitHub account exposed a bug: when `pollMentions` returned 0 items with `since=null`, its `newCursor` calc resolved to `null` → `markPollCompleted` overwrote the cursor that `pollPullRequests` and `pollIssues` had just written. Same shape in slack. Fix is read-before-each-sub-poll — the chain doc's wording ("Reads the current cursor from `integration_state` for each underlying sub-poll") was unambiguous; I misread it the first time. The second cassette test asserts the cursor stays a string across two back-to-back invocations to lock this in.
- **`_unsafeUnwrap`, not literal `throw`.** The chain doc and `.claude/rules/error-handling.md` both call this out: `_unsafeUnwrap()` is a neverthrow API that throws on `Err` (which the wrapping `async` function turns into a rejected promise), but it's NOT a literal `throw` keyword, so the `pnpm cli check-service-boundaries` scanner stays green. The two `poller.ts` files contain `throw` only in JSDoc comments (lines 14 and 13 respectively); no executable `throw` statements.
- **Each integration package now depends on `@slopweaver/mcp-server`** so the `StartSessionPoller` type comes from the canonical source rather than a duplicated structural alias. `boundaries/element-types` allows package → package; there's no cycle because mcp-server doesn't import integration packages back.
- **slack `since` shape.** `pollMentions` / `pollDMs` declare `since?: Date` and the package uses `exactOptionalPropertyTypes`, so `since: undefined` is a compile error. The adapter spreads `{ since: new Date(cursor) } | {}` conditionally instead of passing `since: undefined`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * GitHub: added a public poller for pull requests, issues, and mentions.
  * Slack: added a public poller for mentions and direct messages.

* **Tests**
  * Added cassette-style end-to-end tests validating polling behavior, cursor progression, and idempotency for both integrations.

* **Chores**
  * Declared a shared runtime dependency for both GitHub and Slack integration packages.

[Review Change Stack](https://app.coderabbit.ai/change-stack/slopweaver/slopweaver/pull/47?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by lachiejames** [comment]
**Codex review** — gpt-5.4 (high reasoning) · iteration 1

REVIEW_STATUS: CHANGES_REQUESTED

1. Issue 44's locked scope explicitly requires Polly-cassette tests for both new adapter files, with assertions that invoking the closure produces `evidence_log` rows and advances `integration_state.cursor` ([issue #44](https://github.com/slopweaver/slopweaver/issues/44)). The added tests do not do that. Both `packages/integrations/github/src/poller.test.ts` and `packages/integrations/slack/src/poller.test.ts` explicitly `vi.mock` the underlying pollers and only verify wiring, arg-threading, call order, and rejection behavior. That is a scope miss against the chain doc, not just a different test style.

2. The PR note says \`grep \"throw \" packages/integrations/{github,slack}/src/poller.ts\` returns zero matches, but that check currently fails because both new files still contain \`throw\` in comments. See \`packages/integrations/github/src/poller.ts:14\` and \`packages/integrations/slack/src/poller.ts:13\`. There is no executable \`throw\` statement, but the specific reviewer check and PR claim are still false as written.

Other requested checks passed: the factories use named object params and explicit return types, introduce no \`any\`, keep the required sequential order, do not call \`fetchIdentity\` inside the GitHub adapter, thread \`since\` correctly (\`null\` for GitHub, omitted for Slack), use \`_unsafeUnwrap()\` rather than a literal \`throw\`, add no new \`GithubError\`/\`SlackError\` variants, leave \`StartSessionPoller\` and \`apps/mcp-local/src/cli.ts\` untouched, and stay under the 500-line diff cap excluding the lockfile. The new \`@slopweaver/mcp-server\` dependency is one-way only and does not create a package cycle; the Slack conditional spread is consistent with \`exactOptionalPropertyTypes\`.

File: packages/integrations/github/src/poller.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+export function createGithubPoller({
+  token,
+  username,
+}: CreateGithubPollerArgs): StartSessionPoller {
+  return async ({ db, now }) => {
+    const nowFn = (): number => now;
+
+    const cursorResult = await readCursor({ db, integration: INTEGRATION });
+    const cursor = cursorResult._unsafeUnwrap();
+    const since = cursor ? new Date(cursor) : null;
```
[the remark this answers is not in the document]
Comment (lachiejames) [reply packages/integrations/github/src/poller.ts]: Skipping per `CLAUDE.md`:

> Don't add error handling, fallbacks, or validation for scenarios that can't happen. Trust internal code and framework guarantees. Only validate at system boundaries (user input, external APIs).

The cursor in `integration_state.cursor` is always written by our own `markPollCompleted` from either:
- `items[0].updated_at` — Octokit-typed `string` (or Slack `auth.test` / `search.messages` `ts` → ISO via `slackTsToIso`)
- `since.toISOString()` — always a valid ISO string
- `null`

`new Date(invalidString)` producing `Invalid Date` can only happen if someone hand-edits the SQLite file, which is outside the trust boundary the rule draws. If a future migration introduces a path that writes malformed strings, that's the correct place to add the guard — not at every read site.
